### PR TITLE
Add entrypoint code

### PR DESCRIPTION
**Pull request by dshcherb**
Adds entrypoint code which will be used by charm implementations in
order to initialize the framework and other in-memory state and dispatch
a new Juju event.

### CODE REVIEW

File: test/test_main.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+)
+
+# This relies on the expected repository structure to find
+# a path to the charm under test.
+JUJU_CHARM_DIR = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)),
+    'charms/main'
+)
+
+del sys.path[0]
```
Comment (dshcherb) [inline test/test_main.py]: I tried to avoid a side-effect here but so long as we are using `python3 -m unittest` we cannot use `if __name__ == '__main__'` here to avoid unconditional modification of sys.path. The reason is that, for example, `import juju.main` will try to do `import charm` which depends on having `<JUJU_CHARM_DIR>/lib` in `sys.path` with `<JUJU_CHARM_DIR>/lib/charm.py` available for import.

An alternative to this is using `subprocess` for every test case but it requires some mechanism to communicate charm execution output to the parent process that runs the test.

I am open to suggestions on this one.

File: juju/main.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    # An absolute path is used as using a symlink like
+    # hooks/install -> ../lib/juju/main.py
+    # will result in $JUJU_CHARM_DIR/lib/juju being the first path stored
+    # in sys.path during hook execution which is undesirable.
+    # This happens because the directory containing a script is added to
+    # sys.path by the interpreter which gets resolved if it is a symlink.
+    del sys.path[0]
+    sys.path.insert(0, os.path.join(JUJU_CHARM_DIR, 'lib'))
+
+from charm import Charm # noqa
```
Comment (johnsca) [inline juju/main.py]: I think this is a symptom of not having the separation of concerns and direction of dependency right.  The framework code is expected to be reused across all charms, so it should be independent from the charm and have no dependencies on the charm.  Rather, the framework should be called by the charm and told what implementation to use.  Something like:

```python
import juju.framework

class MyCharm(juju.framework.CharmBase):
    ...

if __name__ == '__main__':
    juju.framework.main(MyCharm)
```

The charm is almost certainly going to want additional third-party libraries at some point, so we'll have to deal with the packaging issue sooner or later.  Once we do so, then it makes sense to me for the framework to be packaged as a library as well (also, components).

With this approach, the path issues are resolved automatically, since Python always knows how to find and import libraries, and the charm code has the right paths because it's the thing that's being executed.

That does leave wiring up at least the initial hook executable to symlink to the charm code, but I think we'll likely end up wanting some manner of build phase regardless, if only to pre-cache the libraries the charm depends on (currently done via the wheelhouse, but there are issues with that approach which need to be addressed) to minimize network traffic during install and handle the restricted network deployment case.  If we presume a build phase, then it seems reasonable to have it do a bit of work to wire up the Python-friendly charm code to the Juju-friendly hook code.

Comment (dshcherb) [reply juju/main.py]: I ended up introducing a couple of things to preserve the current behavior.

If it's a hook execution by a unit agent or a manual hook execution simulation (e.g. in `debug-hooks`), the following will run:

```
if __name__ == '__main__':
    setup_path()
    main()
```

Then in the `main` function I use `importlib` to load the necessary modules which are then used to set hook symlinks up on install/upgrade-charm and instantiate the Charm class.

https://github.com/canonical/charmbase/pull/6/files#diff-460b509ca15cacedde5eeb7bcba28b81R165-R167

As a result, `juju/main.py` can be imported without side-effects and will work as expected and set up `sys.path` in the `__name__ == '__main__'` case. If somebody uses `juju.main` as a module and call `juju.main.main` function, the expectation is that they will have the environment set up correctly.

File: juju/main.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+def _get_charm_dir():
+    charm_dir = os.environ.get("JUJU_CHARM_DIR")
+    if charm_dir is None:
+        # Assume $JUJU_CHARM_DIR/lib/juju/main.py structure.
+        charm_dir = os.path.dirname(os.path.dirname(
+            os.path.dirname(os.path.abspath(__file__))))
+    return charm_dir
+
+
+def _load_metadata():
```
Comment (dshcherb) [inline juju/main.py]: Could move those to `metadata.py` or something else later so that other code could use it (including developer-provided code).

File: juju/main.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+def setup_path():
+    # Use modules from the $JUJU_CHARM_DIR/lib directory (conditionally, to
+    # avoid side-effects on module import).
+    # An absolute path is used as using a symlink like
+    # hooks/install -> ../lib/juju/main.py
+    # will result in $JUJU_CHARM_DIR/lib/juju being the first path stored
+    # in sys.path during hook execution which is undesirable.
+    # This happens because the directory containing a script is added to
+    # sys.path by the interpreter which gets resolved if it is a symlink.
```
Comment (niemeyer) [inline juju/main.py]: I don't quite get the comment above. First, the absolute path is used because it sounds like the right thing to do. If it's relative, it will be relative whereved the current process decided to move over to. And then, it sounds like the comment above is also saying that the directory containing the executable is only being added to the path if it's a symlink, and that wouldn't happen if it was an absolute path, but that's not true I think. I think this comment should have a form closer to:
```
# First element is the directory containing this file. We don't want that in the path.
del ...
# Then, add $JUJU_CHARM_DIR/lib to the path.
... insert ...
```
And done?

File: test/charms/main/lib/charm.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: The filename might be slightly more clear as charms/test_main/... or something like this.. it's reading like "the main charm" now.

File: test/charms/main/lib/charm.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+class HARelationChangedEvent(EventBase):
+    pass
+
+
+class HARelationDepartedEvent(EventBase):
+    pass
+
+
+class HARelationBrokenEvent(EventBase):
+    pass
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: A lot of wasted spacing here. Can we compress this a bit so we can read and reason about more of it at once?

File: test/charms/main/lib/charm.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    db_relation_departed = Event(DBRelationDepartedEvent)
+    db_relation_broken = Event(DBRelationBrokenEvent)
+    mon_relation_joined = Event(MonRelationJoinedEvent)
+    mon_relation_changed = Event(MonRelationChangedEvent)
+    mon_relation_departed = Event(MonRelationDepartedEvent)
+    mon_relation_broken = Event(MonRelationBrokenEvent)
+    ha_relation_joined = Event(HARelationJoinedEvent)
+    ha_relation_changed = Event(HARelationChangedEvent)
+    ha_relation_departed = Event(HARelationDepartedEvent)
+    ha_relation_broken = Event(HARelationBrokenEvent)
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: Looks good for now.. We'll want to use something more dynamic later.

File: test/charms/main/lib/charm.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            self._charm_config = {}
+
+        self._state_file = self._charm_config.get('STATE_FILE')
+        self._state = {}
+        self._state['event_type_names_observed'] = list()
+
+        # __init__ does not use reset_type because the goal is to
+        # avoid those values being reset on every object instantiation.
+
+        charm_events = dict()
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: `{}`

File: test/charms/main/lib/charm.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: Why is this necessary given the version below?

File: test/charms/main/lib/charm.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(self.on).__dict__.items()
+                             if type(instance) == Event})
+
+        event_names = charm_events.keys()
+
+        # Set up handlers default handlers for base events.
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: Sentence isn't reading properly.

File: test/charms/main/lib/charm.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        # __init__ does not use reset_type because the goal is to
+        # avoid those values being reset on every object instantiation.
+
+        charm_events = dict()
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(self.on).__dict__.items()
+                             if type(instance) == Event})
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: And then, why is this being done here?  The point of this test (as I understand it) is ensuring that the entrypoint logic is working correctly for a charm. But instead, this is super specific to the current implementation we have, and not likely to be seen in a normal charm. Instead, I'd expect this to be a much more normal charm, with proper lines that read like
```
framework.observe(self.on.some_event, self.on_some_event)
``` 
For two or three events which are different from each other in some fundamental way (one of them is install, another is a created symlink, another is dynamic, and that's about it?).

File: test/charms/main/lib/charm.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        event_names = charm_events.keys()
+
+        # Set up handlers default handlers for base events.
+        for event_name in event_names:
+            handler_name = f'on_{event_name}'
+            try:
+                getattr(self, handler_name)
+            except AttributeError:
+                setattr(self, handler_name, self._handler_core)
+            self.framework.observe(getattr(self.on, event_name), self)
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: Same as above.. feels super introspective, but it's not clear this is necessary. Am I missing something?

File: test/charms/main/lib/charm.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    def _handler_core(self, event):
+        event_type_name = type(event).__name__
+        logger.debug(f'Handling event {event_type_name}')
+
+        count_called_key = f'{event_type_name}_count_called'
+        count_called = self._state.get(count_called_key)
+        if count_called is None:
+            count_called = 0
+        count_called += 1
+        self._state[count_called_key] = count_called
```
Comment (niemeyer) [inline test/charms/main/lib/charm.py]: Same idea.. it feels very complex for something that a hardcoded couple of events would handle fine.

File: test/test_main.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+
+    def test_all_events_handled(self):
+        # Infer supported events from the charm class and CharmEvents.
+        charm_events = {}
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(charm.Charm.on).__dict__.items()
+                             if type(instance) == Event})
```
Comment (niemeyer) [inline test/test_main.py]: Same comments here to the one in the sample charm.

File: test/test_main.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        logger.debug(f'Expected events {charm_events.keys()}')
+
+        charm_config = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+        }))
+
+        # First run "install" to make sure all hooks are set up.
+        self._simulate_event('install', charm_config)
+
+        # Simulate hook executions for every event.
```
Comment (niemeyer) [inline test/test_main.py]: Simliarly, there's no need for this. This is testing the exact same logic every time.

File: juju/main.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        endpoints = metadata.get(endpoint_type)
+        if endpoints:
+            charm_endpoints.extend(endpoints)
+    return charm_endpoints
+
+
+def _handle_event_link(charm_dir, event_name):
+    event_hook_path = os.path.join(
+        charm_dir,
+        'hooks', event_name.replace('_', '-'))
```
Comment (johnsca) [inline juju/main.py]: I personally find the [pathlib.Path()](https://docs.python.org/3/library/pathlib.html) object-oriented style syntax a bit nicer to work with, but it's just an aesthetic thing.  Just wanted to raise it as an option, but of course whatever we go with, we should be consistent.

File: test/test_main.py (hunk 15 above)
Comment (dshcherb) [reply test/test_main.py]: This is just to make sure that any valid event name is properly handled.

For example, we have:

- single word events: {install, start, ...};
- events with a single dash: {config-changed, upgrade-charm, leader-elected, ...};
- events with multiple dashes: {leader-settings-changed, <endpoint-name>-relation-joined, ...}

I'd skip that if we haven't had to do string manipulation ("-" -> "_" conversion) and dynamic lookup of attributes based on those strings.

All events are tested to make sure there is coverage for anything that might be added to core events in the future.

File: test/test_main.py (hunk 14 above)
Comment (dshcherb) [reply test/test_main.py]: See below.

File: test/charms/main/lib/charm.py (hunk 13 above)
Comment (dshcherb) [reply test/charms/main/lib/charm.py]: See the comment below.

File: test/charms/main/lib/charm.py (hunk 12 above)
Comment (dshcherb) [reply test/charms/main/lib/charm.py]: See the review comment below on event names.

File: juju/main.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            debugf(f'Removing entry {event_hook_path} as it does not point'
+                   ' to "install"')
+            os.unlink(event_hook_path)
+        else:
+            create_link = False
+
+    if create_link:
+        debugf(f'Creating a new symlink named {event_hook_path}'
+               ' to hooks/install')
+        os.symlink('install', event_hook_path)
```
[the remark this answers is not in the document]
Comment (johnsca) [reply juju/main.py]: I don't think this is true, actually.  It's making the symlink target relative to the link (i.e., assuming that all hooks are in the same directory); that is, the link file created (`event_hook_path`) is absolutely referenced, but it points to a file relative to itself (`install`).  That seems ok to me and fairly common practice with symlinks.

File: juju/main.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+    debugf(f'Discovered core charm events: {core_events}')
+
+    # Discover Juju events supported by the juju library
+    # using reflection.
+    supported_juju_events = set()
+    for event_name in dir(Charm.on):
+        attr_type = type(getattr(Charm.on, event_name))
+        if attr_type == BoundEvent and event_name in possible_events:
+            supported_juju_events.add(event_name)
```
Comment (johnsca) [inline juju/main.py]: This can all be dropped in favor of `Charm.on.events()` now that my PR has landed.

File: test/test_main.py (hunk 15 above)
Comment (niemeyer) [reply test/test_main.py]: This is not making sure that all events are tested. It's making sure that the events iterated over are tested, and you have the exact same logic both in the test and in here to iterate over the events. If this list misses half the events, nothing bad will happen. The bug goes unperceived.

Again, the test is simpler, more readable, and more realistic, if the events are hooked up as you expect real people to hook up events.

File: juju/main.py (hunk 16 above)
Comment (niemeyer) [reply juju/main.py]: +1

File: juju/main.py (hunk 18 above)
Comment (niemeyer) [reply juju/main.py]: +1

File: juju/main.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    framework -> charm object -> member child objects (endpoints, components)
+
+    The charm class should hold all necessary setup code for the modeled
+    application such as object instantiation, setup of observation of events
+    (which may also be done in child classes).
+
+    """
+    charm_module = importlib.import_module('charm')
+    juju_charm_module = importlib.import_module('juju.charm')
+    framework_module = importlib.import_module('juju.framework')
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: I thought it would raise PEP8 warnings but apparently not.

File: juju/main.py (hunk 18 above)
Comment (johnsca) [reply juju/main.py]: This makes me wonder if `EventsBase.events()` should in fact return a `dict` rather than a tuple generator.

File: juju/main.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (johnsca) [reply juju/main.py]: Only reason in favor of the current approach that I could see is if we wanted to mock it out to unit test this function.  But I think the consensus was that this function in particular is the one we only ever want to functional test.

[the remark this answers is not in the document]
Comment (johnsca) [reply juju/main.py]: @dshcherb Is the user `charm` module the only one that was causing issues with import paths when done at the top of the module?  If so, then my preference here would be to have the `__main__` section do the user module import and pass the actual `Charm` class into this function, and move the framework imports back up to the top of the module.

File: juju/main.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+    charm_dir = _get_charm_dir()
+    juju_event_name = os.path.basename(sys.argv[0])
+    # Set up hooks on initial install or charm upgrade. Note that if
+    # a charm is force-upgraded this will only be called after
+    # upgrade-charm is executed.
+    if juju_event_name in ['install', 'upgrade-charm']:
```
[the remark this answers is not in the document]
Comment (johnsca) [reply juju/main.py]: I think a set might be even faster than a tuple?  Not certain, though.

[the remark this answers is not in the document]
Comment (niemeyer) [reply juju/main.py]: Haven't measured but I'd be surprised if that was the case. A tuple like this is effectively a couple of pointers and immutable. It's one of the simplest data structures available.

File: test/charms/main/lib/charm.py (hunk 6 above)
Comment (johnsca) [reply test/charms/main/lib/charm.py]: This is actually following the style guidelines for Python.  All linters will reject the `class Foo: pass` style and no blank lines between unless explicitly exempted (which will also affect all other cases where multi-statement lines are used and blank lines missing between class / function definitions).  I'm not sure that we want to go down that road, but maybe there's a linter that we can use which would support exempting only a specific block of code.

We could also get around it with per-line comments, a la:

```python
class InstallEvent(EventBase): pass  # noqa: ignore=E302,E701
class StartEvent(EventBase): pass  # noqa: ignore=E302,E701
class StopEvent(EventBase): pass  # noqa: ignore=E302,E701
class ConfigChangedEvent(EventBase): pass  # noqa: ignore=E302,E701
class UpdateStatusEvent(EventBase): pass  # noqa: ignore=E302,E701
class UpgradeCharmEvent(EventBase): pass  # noqa: ignore=E302,E701
class PreSeriesUpgradeEvent(EventBase): pass  # noqa: ignore=E302,E701
class PostSeriesUpgradeEvent(EventBase): pass  # noqa: ignore=E302,E701
class LeaderElectedEvent(EventBase): pass  # noqa: ignore=E302,E701
class LeaderSettingsChangedEvent(EventBase): pass  # noqa: ignore=E302,E701
```

File: juju/main.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+        else:
+            create_link = False
+
+    if create_link:
+        debugf(f'Creating a new symlink named {event_hook_path}'
+               ' to hooks/install')
+        os.symlink('install', event_hook_path)
+
+
+def _setup_hooks(charm_dir, Charm, CharmEvents, Event, BoundEvent):
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: I'll rebase to Cory's code and rewrite this part because it has a useful generator for this.

File: juju/main.py (hunk 18 above)
Comment (dshcherb) [reply juju/main.py]: Will need this as well: https://github.com/canonical/charmbase/pull/9

Comment (johnsca) [reply juju/main.py]: Updated #9 so this can become:

```python
supported_juju_events = {event_name for event_name in Charm.on.events()
                         if event_name in possible_events}
```

File: juju/main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: I think I'll reword the message to make it clearer:

```
        debugf(f'Creating a new relative symlink at {event_hook_path}'
               f' to poiting to "install" located in {charm_dir}/hooks')
```
The relative symlink does point to "install" but the location of the symlink itself is also useful to know.

File: juju/main.py (hunk 16 above)
Comment (dshcherb) [reply juju/main.py]: OK, I'll change it.

File: juju/main.py (hunk 4 above)
Comment (dshcherb) [reply juju/main.py]: Agreed, I think we can have a much shorter comment.

File: test/test_main.py (hunk 15 above)
Comment (dshcherb) [reply test/test_main.py]: OK, agreed.

File: juju/main.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+    The expected hierarchy is a single charm object under the framework
+    object with many child objects under the charm object:
+
+    framework -> charm object -> member child objects (endpoints, components)
+
+    The charm class should hold all necessary setup code for the modeled
+    application such as object instantiation, setup of observation of events
+    (which may also be done in child classes).
+
+    """
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: OK, we can place something like that into developer docs.

File: juju/main.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: @johnsca 

`sys.path` will not contain `lib` at the beginning of execution so `charm` is not the only thing - all `juju.` imports will fail as well.

File: test/charms/main/lib/charm.py (hunk 7 above)
Comment (dshcherb) [reply test/charms/main/lib/charm.py]: Removed it as Cory's PR was merged.

File: test/charms/main/lib/charm.py (hunk 6 above)
Comment (dshcherb) [reply test/charms/main/lib/charm.py]: Got rid of this because Cory's PR got merged.

File: test/charms/main/lib/charm.py (hunk 5 above)
Comment (dshcherb) [reply test/charms/main/lib/charm.py]: OK, I'll change it.

**Comment by johnsca** [comment]
Odd that Travis passed it on 3.7 but it failed on 3.6

File: juju/main.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+
+CHARM_STATE_FILE = '.unit-state.db'
+
+
+def debugf(format, *args, **kwargs):
+    pass
+
+
+class InvalidHookEntry(Exception):
+    pass
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: Not anymore, I was using it previously to handle directories named after hooks (e.g. in case a developer erroneously places a "start" or "config-changed" directory under hooks directory). The current code will just throw an IsADirectoryError while trying to unlink a directory.

File: juju/main.py [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+
+    charm_dir -- A root directory of the charm.
+    Charm -- A type of a concrete charm implementation.
+    CharmEvents -- A type declaring core Juju events.
+    Event -- A framework Event type.
+    BoundEvent -- A framework BoundEvent type.
+    """
+    for event_name in charm.on.events().keys():
+        if event_name != 'install':
+            _handle_event_link(charm_dir, event_name)
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: Agreed.

File: juju/main.py [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+    # operation.
+    charm_state_path = charm_dir / CHARM_STATE_FILE
+    framework = juju.framework.Framework(data_path=charm_state_path)
+    try:
+        # Set up charm-specific in-memory state.
+        charm = charm_module.Charm(framework, None, _load_metadata())
+
+        # Set up hooks on initial install or charm upgrade. Note that if
+        # a charm is force-upgraded this will only be called after
+        # upgrade-charm is executed.
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: It only gets run if a unit is not in the error state, otherwise Juju just waits until you resolve a unit and then it runs the failed hook again followed by `config-changed`.

https://gist.github.com/dshcherb/c828894bd385ac4662e227d93c0ddcb7

I can make the comment more specific in saying that force-upgrade on a unit in the error state will not result in symlinks getting recreated.

The problem with checking for a revision is that I have to store state somewhere locally to compare against.

I could add config-changed to the list which will make it a more expensive operation as it will run more frequently. However, I don't think it's nearly as bad as executing it for relation events because config-changed is operator-initiated in most cases.

File: juju/main.py [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+    Hooks are created as symlinks to "install" which may be either
+    a copy of main.py (which calls the main function) or a symlink
+    to it. Note that it is important to avoid creating a recursive
+    symlink which is why the install event itself is skipped.
+
+    charm_dir -- A root directory of the charm.
+    Charm -- A type of a concrete charm implementation.
+    CharmEvents -- A type declaring core Juju events.
+    Event -- A framework Event type.
+    BoundEvent -- A framework BoundEvent type.
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: Thanks.

File: juju/main.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: Added config-changed to the new revision.

[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: I think storage hooks will need to be handled as well.

Judging by the code and public documentation it looks like "install" will run after any storage attachment events.

https://github.com/juju/juju/blob/692c32713224c80022aec33da220b3525efa582f/worker/uniter/storage/resolver.go#L78-L82

https://jaas.ai/docs/writing-charms-that-use-storage#heading--storage-hooks
> [name]-storage-attached hooks will be run before the install hook

[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: But the challenge with storage events is that hooks for them have to be there at the deployment time e.g.:

disks-storage-attached -> install

And the first event a unit agent will run will be `hooks/disks-storage-attached`, not `hooks/install`.

File: juju/main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply juju/main.py]: Yeah, I was wrong, sorry. I read the logic reversed.

File: juju/main.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: I added a symlink for a storage endpoint called "disks" to the test charm and updated the `setup_hooks` test to run a storage event before the install event.

So, for now, the assumption is that a charm author will create an install symlink to `../lib/main.py` and any storage hook symlinks pointing to the `install` symlink.

File: test/charms/test_main/lib/charm.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+        count_called += 1
+        self._state[count_called_key] = count_called
+
+        count_deferred_key = f'{event_type_name}_count_deferred'
+        count_deferred = self._state.get(count_deferred_key)
+        if count_deferred is None:
+            count_deferred = 0
+
+        if deferred:
+            count_deferred += 1
```
Comment (niemeyer) [inline test/charms/test_main/lib/charm.py]: Still a lot of generic logic in this function which might be replaced with simpler, more readable, and more reliable logic (in terms of testing) if hardcoded.

I would imagine something along those lines for example:
```
def __init__(self):
    ...
    self._state["on_install"] = []

def on_install(self, event):
    self._state["on_install"].append(event.event_type)
```

Despite being much simpler, note that this test will actually allow detecting if we call on_install with the wrong event, while the current logic doesn't.

File: test/test_charm.py [hunk 28]
```diff
@@ -15,9 +17,14 @@ class TestCharm(unittest.TestCase):
     def setUp(self):
         self.tmpdir = Path(tempfile.mkdtemp())
 
+        # Avoid introducing side-effects for other tests by using a separate type.
+        CharmBase.on = types.new_class('TestCharmEvents', bases=(CharmEvents,))()
```
Comment (niemeyer) [inline test/test_charm.py]: This can be normal Python code:
```
class TestCharmEvents(CharmEvents): pass
```

File: test/test_charm.py [hunk 29]
```diff
@@ -15,9 +17,14 @@ class TestCharm(unittest.TestCase):
     def setUp(self):
         self.tmpdir = Path(tempfile.mkdtemp())
 
+        # Avoid introducing side-effects for other tests by using a separate type.
```
Comment (niemeyer) [inline test/test_charm.py]: This should be more clear about which side effects it's talking about, otherwise we'll have to go after boring details to figure it out again in a few months:
```
# Relations events are defined dynamically and modify the class attributes.
# We use a subclass temporarily to prevent these side effects from leaking. 
```

File: juju/main.py [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+    Hooks are created as symlinks to "install" which may be either
+    a copy of main.py (which calls the main function) or a symlink
+    to it. Note that it is important to avoid creating a recursive
+    symlink which is why the install event itself is skipped.
+
+    charm_dir -- A root directory of the charm.
+    charm -- An instance of the Charm class.
+    """
+    for event_name in charm.on.events().keys():
+        if event_name != 'install':
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: Agreed, it feels like `_handle_event_link` should be the one checking for this.

File: juju/main.py [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+        debugf(f'Emitting Juju event {event_name}')
+        event_to_emit.emit()
+
+
+def main():
+    """Setup the charm and dispatch the observed event.
+
+    The event name is based on the way this executable was called (argv[0]).
+    """
+    import charm as charm_module
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: If somebody ships an abstract charm as a module I guess they can use a different type name for it for others to create their `class Charm(AbstractCharm)`. Even then it can be imported as something else to be used for inheritance.

If we have multiple tests though that use different modules imported as `charm` for the module name then we will have to unload this module on `tearDown`.

File: test/charms/test_main/lib/charm.py (hunk 27 above)
Comment (dshcherb) [reply test/charms/test_main/lib/charm.py]: I simplified the test according to your feedback.

I still need to have a list of observed events (per hook invocation/Charm object lifetime) to check for the order of event handling as seen by the charm.

`        self._state['observed_event_types'] = []`

Otherwise, my test code would have introduced its own ordering by reading per-handler states one by one.

File: juju/main.py [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+    # TODO: If Juju unit agent crashes after exit(0) from the charm code
+    # the framework will commit the snapshot but Juju will not commit its
+    # operation.
+    charm_state_path = charm_dir / CHARM_STATE_FILE
+    framework = juju.framework.Framework(data_path=charm_state_path)
+    try:
+        # Set up charm-specific in-memory state.
+        charm = charm_module.Charm(framework, None, _load_metadata())
+
+        # Set up hooks on initial install or charm upgrade. Note that if a charm is force-upgraded,
```
[the remark this answers is not in the document]
Comment (dshcherb) [reply juju/main.py]: The comment as a whole isn't outdated, config-changed is mentioned as a part of the second sentence with regards to forced charm upgrade. I don't mind rewording it though.

File: test/test_charm.py (hunk 28 above)
Comment (johnsca) [reply test/test_charm.py]: I originally suggested `new_class` because I assumed that the `TestCharmEvents` class definition would be persistent even inside a function call.  But I tested it and realized that assumption was wrong, which makes sense now that I think about it again.  Good to know, and this is much more simple.
